### ISSUE TITLE
Collapsing the drawer and reopening to partiallyRevealed disables button

### ISSUE THREAD

**Issue by daniel-choi** [52inc/Pulley#170]
So I have a UIButton inside UIView and this view is hidden when the drawer is collapsed. The button works fine when you first open the drawer. However, after you collapse the drawer and open back to ```.partiallyRevealed``` the button doesn't respond to the touch anymore (My rough guess is userInteraction gets disabled?).
Interesting part is this only happens when ```supportedDrawerPositions``` is ```[.partiallyRevealed, .collapsed]``` It works fine when it's set to ```.all``` or ```[.partiallyRevealed, .collapsed, .open]```.
Do you have any idea why this is happening?
Thank you!

**Comment by brendan09** [comment]
I don't have any immediate thoughts on why this would happen. The Sample project doesn't exhibit this issue with it's tableview, so that means that the problem is most likely somewhere in your project external to Pulley. The only lines in Pulley that adjust the isUserInteractionEnabled setting on a view is 3 lines that deal with the background dimming view that's behind the drawer. This wouldn't affect your button, however.

**Comment by brendan09** [comment]
When you change `. partialRevealDrawerHeight` make sure you immediately call `setNeedsSupportedDrawerPositionsUpdate()` to update the internal layout. If you don't then it's possible your button is actually _outside_ the bounds of one of the container views in Pulley. If `clipsToBounds` were set to true on that container view then you wouldn't see your button at all (or it would be cut in half, etc.)...but since the default is false you'll see it, just not be able to interact with it.

**Comment by daniel-choi** [comment]
Sorry, maybe I wrote that wrongly. ```partialRevealDrawerHeight``` is set using the ```PulleyDrawerViewControllerDelegate```. 
Here is the code for the delegate:
``` swift
func partialRevealDrawerHeight(bottomSafeArea: CGFloat) -> CGFloat {
        return contentView.bounds.height + bottomSafeArea + offset // offset = 4
}
```
```contentView``` here is the UIView wrapping all the contents for drawer to render in ```.partialReveled``` mode.

**Comment by brendan09** [comment]
Yes, that's what I'm referring to. That can't be a dynamic value. It has to be a specific number, and if it changes you need to call the function I mentioned above.

What is returned here (if you're not using the .open position) is factored in to the total height of your drawer view controller. If you give it a small number (Example: 100) and the suddenly it's now 200 when the user drags the drawer...but Pulley doesn't know you've changed it...you'll be able to drag it higher than what the layout is prepared for. As a result, the internal drawer container view will 'clip' your drawerViewController's view. Your button is most likely out of bounds (although clipping is disabled, so you can still visibly see it) because Pulley didn't know you made that position larger. You must tell Pulley when that value changes.

Your next question is probably "If that's the case, why doesn't Pulley have this as a property that can update internals as needed instead of a delegate method?". It's setup this way so that the drawer view controller can be easily changed (and animated) without needing a half dozen parameters to the 'setDrawerViewController' method, and requiring extensive parameter lists anytime anything is changed. There are some other reasons around architecting ownership of those properties (and how it's more a property of the drawer than Pulley itself), but that's the big user-facing reason they're done as delegate methods.

**Comment by brendan09** [comment]
Its similar in concept to the way updating Auto Layout constraints work: You're welcome to make changes to constraints, but if you want them to properly be reflected in the layout then you have to tell the system to make a layout pass with layoutIfNeeded, otherwise your layout isn't current and you can see weird side-effects when you try to interact with it until it's consistent again.

**Comment by daniel-choi** [comment]
Thank you for your thorough answer! I understand why now. One more question... Where can I put `setNeedsSupportedDrawerPositionsUpdate()`? I've put it on `func drawerPositionDidChange()` but that didn't work and I have nowhere else I can call this method which comes after `partialRevealDrawerHeight()` .

**Comment by brendan09** [comment]
> comes after func partialRevealDrawerHeight()

Why do you need it to come _after_ that? Why are you trying to have a dynamic height here? Can it not be pre-computed and stored, then updated as needed?

**Comment by daniel-choi** [comment]
Okay, so I calculated height on `viewDidLayoutSubviews` and used that value in `partialRevealDrawerHeight()` and the problem was still there even if I don't change the height dynamically. After few debugging, I found that if I cast the value to `Int` and back to `CGFloat` fixes the problem. I have no clue why this is happening. The original value I got was `CGFloat: 262.33333333333337` and below is what I did:

``` swift
func partialRevealDrawerHeight(bottomSafeArea: CGFloat) -> CGFloat {
        return CGFloat(Int(contentView.bounds.height)) + bottomSafeArea + offset // offset = 4
}
```
